Fork master flexibility.In connection with the bootstrap pull request(#6): NEW PULL REQUEST

gvissers13 opened the pull request:
All my modifications concerning the software router:
Included: - migration new Java Client
-migration latest version of RabbitMQ Client
-Boutique's tag
-Improvement of flexibility
-cleaning
-....

gvissers13 committed:
Adding some log for the method createSession(). Useful to debug a connection

gvissers13 committed:
Modification: add methods to retrieve boutique's tags like. It'll be useful for later(For example, for lauching an algo on the proper server)

gvissers13 committed:
Modification: add methods to retrieve boutique's tags. For now, it's on comments... But, when the new architecture will be up, we'll be able to uncomment

gvissers13 committed:
"BUG" Discovered: add a comment and a log line to understand why the application can be blocked

gvissers13 committed:
use of the RabbitMQ version 5.6.0 java client.
In this version, the "QueuingConsumer" class is removed.
As a result, we had to change the way messages are processed.
2 classes will be add later: RabbitMQConsumerComThread and RabbitMQConsumerProcessingServer.

gvissers13 committed:
-add 2 new classes... Because we change the RabbitMQ client. So, it's the new way to implement a consumer
-Use of the new Java client

gvissers13 committed:
use of the new java client

gvissers13 committed:
clean: there was an "extra" "s".... Careless mistake

gvissers13 committed:
use of the new java client in the class Main

gvissers13 committed:
add the queueCommunicationRetrieve. Useful for the new architecture( where there is a retrieve queue between the SR and the Core. It's useful because the SR will be able to send information to the Core without API

gvissers13 committed:
Add a flexible way to manage variable

urubens commented on src/main/groovy/be/cytomine/software/repository/SoftwareManager.groovy:
```diff
@@ -67,7 +69,9 @@ class SoftwareManager {
-                            Main.cytomine.deprecateSoftware(currentSoftware.getId())
+                            Software software= new Software()
+                            software.set("deprecated",true)
+                            software.save()
```
Have you tested this ?
It can't work.

urubens commented on src/main/groovy/be/cytomine/software/consumer/threads/JobExecutionThread.groovy:
```diff
@@ -83,24 +85,25 @@ class JobExecutionThread implements Runnable {
             }
             log.info("${logPrefix()} The job [${cytomineJobId}] has been killed successfully !")
-            Main.cytomine.changeStatus(cytomineJobId, 8, 0) // Cytomine.JobStatus.KILLED = 8
+            job.changeStatus(cytomineJobId,8, 0)
```
Why not JobStatus.KILLED ?

urubens commented on src/main/groovy/be/cytomine/software/boutiques/Interpreter.groovy:
```diff
@@ -78,6 +78,36 @@ class Interpreter {
+    }
+
+    def getDiskSpaceNeeded(){
+        def numberOfDiskSpace=descriptor?."Disk-space"
```
Keep same nomenclature as other "Disk-space" vs "disk-space" ...

gvissers13 committed:
Changes due to the pull request: modification of the boutique tag. Now, the tag disk-space as the same nomenclature as others

gvissers13 committed:
Changes due to the pull request: Instead of using value, we'll use "JobStatus.KILLED"

gvissers13 committed:
Changes pull request: careless mistakes... Usage of save instead of update.

gvissers13 committed:
Add a new SSH method: execute a command without create a new session
Create a new session cost time... For execute a command on a server, it's okay... But if we want to execute few commands, it becomes problematic. A solution is to keep alive the session and to execute commands on this session. This is the purpose of the new method executeCommandWithoutCreateNewSession.

gvissers13 committed:
Add a new class. This class will bring some features. basically, the class will bring tools to check the load of a processing server.
So, we've a method which will retrieve the names of all the partitions. A method which will retrieve the informations about each partitions of a given PS. A method which will do the same but for Nodes.
2 method which will manipulate JSon object ( JSon to lists and lists to JSon).
a display method.

gvissers13 committed:
-add a method for initiate a ssh connection( class: CheckingLoadSlurmProcessingServer)
Thanks to that, we'll be able to retrieve all the informations for the currentPS
-add a method to retrieve all the infos(nodes and partitions) and return a JSon object of them

Some cleaning( logs,comments,...)

gvissers13 committed:
New class which will contains algorithms. Algorithms for choosing the best ProcessingServer
Currently, there's only an Algo. This algo doesn't work yet...
He will just return the id of the "good" processing server

gvissers13 committed:
Add a new method which will return the most suitable PS.
Actually, this method will get, for each processing server, all the load's information. Then, this method will send these info to an algorithm ( from the algo class)
and then, will return the chosenPS

gvissers13 committed:
add a case in the communicationThread switch: This case will manage the workflows..
Basically, when a user want to execute a job, the execute request will pass here before. Then, the com thread will redirect the request.
Here, we've 2 "case". Either, we want an automatic choice: thanks to an algo, the SR will choose the best suitable PS
Or we doesn't want that... The SR will simply redirect the request

gvissers13 committed:
Add a new case: this case will manage the check load request for a given ps
So, basically, we'll simply check the load with the getFullInformation method and then, we'll publish the result on the communicationRetrieveQueue

gvissers13 committed:
Add the com.google.guava + change the version of the rabbitmqClient

gvissers13 committed:
modification of the "retrieve fields nodes" algorithm. before, it didn't work for some cases... Like, if we had several nodes for example.
Now, it works for every case

gvissers13 committed:
add a new algorithm: random algorithm which will choose a valid PS with a random number

gvissers13 committed:
fixing the crashes during the Periodical check of software repositories...
The crashe ocurred because the github API limit was reached. With an anonymously connection, the "rate limit" is 60... When we exceed this limit, the API return an error. In the current version of the Software router, we don't "handle" that kind of errors... A possible solution could be to use this:
try{
GitHub gitHub=new GitHubBuilder().withRateLimitHandler(RateLimitHandler.FAIL).build()
ghUser = gitHub.getUser(username)
}
catch(IOException e)
{
log.info(e.printStackTrace())
}

As a result, if an exceed limit error appear, the code will throw an IO exception and the thread won't be blocked
NB: we've to surround each use of the API with a try/catch block

gvissers13 committed:
Modification of the management of the "refreshRepositories" request.
Now, we'll answer to the core thanks to the retrieveQueue. We'll send a JSON object with a "response" property. (ok/nok)

the method refreshAll will return ok/nok. ok if the refresh didn't fail and nok in the others case

gvissers13 committed:
Improvement of status's job update. Before, the SR didn't manage the running status. It's was the software that did that!
now, for a better flexibility, the SR will put the status running automatically. Soon as the job is launched

gvissers13 committed:
Correction of a bug. This bug concerned a condition. It missed a "!"

gvissers13 committed:
Adding for the new SSH architecture. Concern the Path of the keyfile.

gvissers13 committed:
some cleaning...

gvissers13 committed:
adding a method that check if a processing server is valid or not
A ps is valid if we can ping it and if singularity is installed on it.

We add an exitcode in the ssh class. This exitcode will provide to us an usefull information about the execution of a command via SSH

gvissers13 committed:
slightly modification: use of a Collection of ParameterConstraint  instead of a "ParameterConstraintCollection"

gvissers13 committed:
Adding some logs and a different way to instanciate the Job

Now, the SR will change the status of the job when this one will be launched

gvissers13 committed:
Adding a new condition for the GitHubManager... Check if the GithubUser isn't null

gvissers13 committed:
little cleaning

gvissers13 committed:
slightly modifications: resolution of a little bug( during the change of the status...We tried to retrieve the status on a null job).
Another resolution of a bug: the parameter "domain class" in the Attachedfile constructor was wrong. This error caused a bug in the Core

gvissers13 committed:
Cleaning+ optimization

urubens commented on src/main/groovy/be/cytomine/software/communication/SSH.groovy:
```diff
@@ -44,22 +55,28 @@ class SSH implements Communication {
     @Override
     def executeCommand(String command) throws JSchException, UnknownHostException {
-        Session session = createSession()
+        session = createSession()
```
So you create session in constructor and re-create it here. 
Please make something efficient. Static method ? Same for other methods...

urubens commented on src/main/groovy/be/cytomine/software/consumer/threads/ProcessingServerThread.groovy:
```diff
@@ -43,18 +43,16 @@ class ProcessingServerThread implements Runnable {
-                    keyFilePath: Main.configFile.cytomine.software.sshKeysFile
-            )
+            String keyFilePath=Main.configFile.cytomine.software.sshKeysFile
+            def keyPathToReturn = """${keyFilePath}/${processingServer.getStr("host")}/${processingServer.getStr("host")}"""
```
keyPath or something like this is not ok ? Why "ToReturn", you return nothing here

urubens commented on pom.xml:
```diff
@@ -60,7 +60,12 @@
+            <version>5.6.0</version>
+        </dependency>
+        <dependency>
+            <groupId>com.google.guava</groupId>
```
Where is it used ?

urubens commented on src/main/groovy/be/cytomine/software/CheckingLoadSlurmProcessingServer.groovy:
```diff
@@ -0,0 +1,259 @@
+import org.json.simple.JSONObject
+
+@Log4j
+class CheckingLoadSlurmProcessingServer {
```
As it's supposed to be generic with respect to a processing method, where is the base class ?

urubens commented on src/main/groovy/be/cytomine/software/ProcessingServerSelectionAlgorithms.groovy:
```diff
@@ -0,0 +1,22 @@
+import be.cytomine.client.models.ProcessingServer
+import org.json.simple.JSONObject
+
+class ProcessingServerSelectionAlgorithms {
```
Don't you think a class hierarchy is more suitable ?

urubens reviewed (changes requested):
I didn't check everything especially long methods, I suppose you have tested them.

gvissers13 committed:
Pull request changes requested: Modification of the Algorithm hierarchy.
Now, we've an abstract class that manage selection's algorithms. To create a new algorithm of selection, we simply have to add a new class that extends this abstract class.

gvissers13 committed:
Pull request change requested: efficient way to generate ssh session.
More efficient than the previous version ( creation of the session was done in the constructor AND in some methods.... )

gvissers13 committed:
Pull request change requested: cleaning for the keypath+ some cleanings

gvissers13 committed:
Pull request change requested: changes about the generic problems (checkLoad only for the slurmMethod)
now, we're generic! so, as a result, we can check the load for a slurm server or another server,....
The processingMethod will implement the way to check the load of the server. (it will depend of the context... The instance of this class actually)

gvissers13 committed:
Update Dockerfile

gvissers13 committed:
Adding an API endpoint in the SR side: usefull to retrieve the load of all processing Server

gvissers13 committed:
Cleaning + adding logs for a better understanding

gvissers13 committed:
Merge remote-tracking branch 'fork-master/cleaning-git' into cleaning-git
